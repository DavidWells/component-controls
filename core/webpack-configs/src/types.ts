import { Configuration } from 'webpack';

export interface RuleOptions {
  name: string;
  config: Configuration;
}
export type RuleType = string | RuleOptions;

export type RuleTypes = RuleType[];
