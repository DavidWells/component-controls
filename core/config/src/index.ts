import * as path from 'path';
import * as fs from 'fs';
import { sync as globSync } from 'glob';
import yargs from 'yargs';
import { Configuration } from '@component-controls/specification';

export const configFileNames = ['main.js', 'main.json', 'main.ts'];

export interface ConfigrationResult {
  config: Configuration;
  configPath: string;
}
/**
 * return the configration object
 * command line accepts -c/ -config parameter for config path
 * the config file is assumed named main.js/main.ts
 * @param baseFolder project folder to start the searh with
 */
export const getConfiguration = (
  baseFolder: string,
  args: string[] = process.argv,
): ConfigrationResult | undefined => {
  const argv = yargs(args)
    .option('config', {
      alias: 'c',
      description: 'configuration folder',
      type: 'string',
    })
    .help()
    .alias('help', 'h').argv;

  const configPath = argv.config
    ? path.resolve(baseFolder, argv.config)
    : baseFolder;
  const configFiles = fs.readdirSync(configPath);
  const configFile = configFiles.find(file =>
    configFileNames.includes(file.toLowerCase()),
  );

  return configFile
    ? {
        config: require(path.resolve(configPath, configFile)),
        configPath,
      }
    : undefined;
};

/**
 * find the story files out of a configuration file
 * using glob for the regex file search
 * @param param0 configuration info return by getConfiguration
 */
export const extractStories = ({
  config,
  configPath,
}: ConfigrationResult): string[] | undefined => {
  const stories =
    config && config.stories
      ? config.stories.reduce((acc: string[], storyRg: string) => {
          const regex = storyRg.replace(/\((.*?)\)/g, '+($1)');
          return [...acc, ...globSync(path.resolve(configPath, regex))];
        }, [])
      : undefined;
  return stories;
};
