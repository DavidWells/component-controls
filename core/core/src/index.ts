export * from './utils';
export { randomizeData } from './randomizeData';
export * from './smartControls';
export * from './source';
export * from './faker';
