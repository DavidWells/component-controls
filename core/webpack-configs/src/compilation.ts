import path from 'path';
import { Compiler, Stats, Configuration } from 'webpack';
import { LoadingStore } from '@component-controls/store';
import { LogOptions } from '@component-controls/logger';
import { RuleTypes } from './types';

/**
 * configuration properties for compile and run
 */
export interface CompileProps {
  /**
   * webpack configuration object
   */
  webPack?: Configuration;
  /**
   * a list of webpack configuration presets from webpack-configs packages
   */
  presets?: RuleTypes;
  /**
   * path to the configuration file e.g : '.storybook'
   */
  configPath?: string;

  /** public output folder for the bundle */
  distFolder?: string;

  /** public file name the bundle, by default 'component-controls.js' */
  bundleName?: string;

  /** public output folder for the assets like images */
  staticFolder?: string;

  /**
   * logger options
   */
  logOptions?: Partial<LogOptions>;
}

const defaultPresets = ['react', 'react-docgen-typescript'];

export const defaultCompileProps: CompileProps = {
  presets: defaultPresets,
};
/**
 * return type from compile and watch functions
 */
export interface CompileResults {
  /**
   * bundle full path and name
   */
  bundleName: string;
  /**
   * the webpack stats object
   */
  stats: Stats;
  /**
   * the stories store
   */
  store?: LoadingStore;
}

export type WatchOptions = Parameters<Compiler['watch']>[0];

/**
 * adds webpack WatchOptions to the Compiler options
 */
export type WatchProps = {
  watchOptions?: WatchOptions;
} & CompileProps;

export const defBundleName = 'component-controls.js';

export const getBundleName = (options: CompileProps) =>
  path.join(
    options.distFolder || `${path.join(process.cwd(), 'public')}`,
    options.bundleName || defBundleName,
  );
