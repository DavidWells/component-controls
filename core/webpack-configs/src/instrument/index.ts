import { Configuration } from 'webpack';

export const instrument: Configuration = {
  module: {
    rules: [
      {
        test: /\.(ts|tsx)$/,
        use: [
          {
            loader: require.resolve('babel-loader'),
            options: {
              presets: [['react-app', { flow: false, typescript: true }]],
            },
          },
        ],
      },
      {
        test: /\.(md|mdx)$/,
        exclude: [/node_modules/],
        loader: 'babel-loader',
        options: {
          presets: [
            [require.resolve('@babel/preset-env'), { modules: 'commonjs' }],
            require.resolve('@babel/preset-react'),
          ],
        },
      },
      {
        test: /\.(md|mdx)$/,
        exclude: [/node_modules/],
        loader: '@component-controls/loader/loader',
        options: {
          mdx: {
            transformMDX: true,
          },
        },
      },
      {
        test: /\.(story|stories).(js|jsx|ts|tsx)$/,
        loader: '@component-controls/loader/loader',
        exclude: [/node_modules/],
        enforce: 'pre',
      },
    ],
  },
  resolve: {
    extensions: ['.ts', '.tsx', '.js', '.jsx'],
  },
};
