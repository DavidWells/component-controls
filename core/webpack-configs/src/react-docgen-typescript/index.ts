import { Configuration } from 'webpack';

export const reactDocgenTypescript: Configuration = {
  module: {
    rules: [
      {
        test: /\.(story|stories).(js|jsx|ts|tsx|mdx)$/,
        loader: '@component-controls/loader/loader',
        exclude: [/node_modules/],
        enforce: 'pre',
        options: {
          propsLoaders: [
            {
              name: '@component-controls/react-docgen-info',
              test: /\.(js|jsx)$/,
            },
            {
              name: '@component-controls/react-docgen-typescript-info',
              test: /\.(ts|tsx)$/,
            },
          ],
        },
      },
    ],
  },
};
