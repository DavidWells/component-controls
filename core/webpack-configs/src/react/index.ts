import { Configuration } from 'webpack';

export const react: Configuration = {
  performance: { hints: false },
  module: {
    rules: [
      {
        test: /\.(js|jsx)$/,
        exclude: [/node_modules/],
        use: [
          {
            loader: 'babel-loader',
            options: {
              presets: [
                [require.resolve('@babel/preset-env'), { modules: 'commonjs' }],
                require.resolve('@babel/preset-react'),
              ],
            },
          },
        ],
      },
      {
        test: /\.(ts|tsx)$/,
        exclude: [/node_modules/],
        use: [
          {
            loader: require.resolve('babel-loader'),
            options: {
              presets: [['react-app', { flow: false, typescript: true }]],
            },
          },
        ],
      },
      {
        test: /\.mdx$/,
        exclude: [/node_modules/],
        use: [
          {
            loader: 'babel-loader',
            options: {
              presets: [
                [require.resolve('@babel/preset-env'), { modules: 'commonjs' }],
                require.resolve('@babel/preset-react'),
              ],
            },
          },
          {
            loader: '@component-controls/loader/loader',
            options: {
              mdx: {
                transformMDX: true,
              },
            },
          },
        ],
      },
      {
        test: /\.(story|stories).(js|jsx|ts|tsx)$/,
        loader: '@component-controls/loader/loader',
        exclude: [/node_modules/],
        options: {
          mdx: {
            transformMDX: true,
          },
        },
      },
    ],
  },
  resolve: {
    extensions: ['.ts', '.tsx', '.js', '.jsx'],
  },
};
