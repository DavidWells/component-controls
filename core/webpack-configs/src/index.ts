const merge = require('deepmerge');
import { Configuration } from 'webpack';
import { react } from './react';
import { instrument } from './instrument';
import { reactDocgen } from './react-docgen';
import { reactDocgenTypescript } from './react-docgen-typescript';
import { RuleOptions, RuleTypes, RuleType } from './types';
export * from './types';

export const presetsFactory: {
  [key: string]: Configuration;
} = {
  'react-docgen-typescript': reactDocgenTypescript,
  'react-docgen': reactDocgen,
  instrument,
  react,
};

const arrayMerge = (dest: any[], src: any[]) => {
  const srcCache = src ? [...src] : [];
  const destItems = dest
    ? dest.reduce((acc, item) => {
        // merge presets by loader or test
        const srcItemIdx = srcCache.findIndex(
          d => d.test === item.test || d.loader === item.loader,
        );
        if (srcItemIdx >= 0) {
          const srcItem = srcCache[srcItemIdx];
          srcCache.splice(srcItemIdx, 1);
          return [...acc, deepMerge(srcItem, item)];
        }
        return [...acc, item];
      }, [])
    : src || [];
  return [...destItems, ...srcCache];
};

const deepMergeWithPresets = (dest: any, source: any) => {
  return dest && source
    ? merge(dest, source, {
        arrayMerge: arrayMerge,
      })
    : source || dest || {};
};

const deepMerge = (dest: any, source: any) => {
  return dest && source
    ? merge(dest, source, {
        arrayMerge: (d: any[], s: any[]) => (d && s ? [...d, ...s] : d || s),
      })
    : source || dest || {};
};
/**
 * expands the presets into webpack config
 * @param presets custom config
 */
export const getWebpackConfig = (presets: RuleTypes): Configuration => {
  const result: Configuration = presets.reduce(
    (acc: Configuration, config: RuleType) => {
      if (typeof config === 'string') {
        const f = presetsFactory[config];
        return deepMergeWithPresets(acc, f);
      }
      if (config && (config as RuleOptions).name) {
        const name = (config as RuleOptions).name;
        if (presetsFactory[name]) {
          const r: Configuration = presetsFactory[name];
          const o: Configuration = (config as RuleOptions).config;
          const merged: Configuration[] = deepMergeWithPresets(r, o);
          return deepMergeWithPresets(acc, merged);
        }
      }
      return deepMergeWithPresets(acc, config);
    },
    {},
  );
  return result;
};

/**
 * merge webpack config with custom set of presets
 * @param webPack passed configuration to merge with
 * @param presets custom config
 */

export const mergeWebpackConfig = (
  webPack?: Configuration,
  presets?: RuleTypes,
): Configuration => {
  if (!presets) {
    return {};
  }
  const newConfig = getWebpackConfig(presets);
  return deepMerge(webPack || {}, newConfig);
};
