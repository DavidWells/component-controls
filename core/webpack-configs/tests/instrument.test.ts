import { getWebpackConfig } from '../src/index';

describe('merge instrument and react-docgen-typescript', () => {
  it('get config', () => {
    expect(
      getWebpackConfig(['instrument', 'react-docgen-typescript'], {}),
    ).toMatchSnapshot();
  });
  it('merge options', () => {
    expect(
      getWebpackConfig(
        [
          'instrument',
          {
            name: 'react-docgen-typescript',
            config: {
              module: {
                rules: [
                  {
                    loader: '@component-controls/loader/loader',
                    options: {
                      //instrumentation options
                      prettier: {
                        tabWidth: 4,
                      },
                    },
                  },
                ],
              },
            },
          },
        ],
        {},
      ),
    ).toMatchSnapshot();
  });
});
