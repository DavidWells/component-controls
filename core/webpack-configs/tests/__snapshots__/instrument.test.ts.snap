// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merge instrument and react-docgen-typescript get config 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "test": /\\\\\\.\\(ts\\|tsx\\)\\$/,
        "use": Array [
          Object {
            "loader": "/Users/atanasster/component-controls/node_modules/babel-loader/lib/index.js",
            "options": Object {
              "presets": Array [
                Array [
                  "react-app",
                  Object {
                    "flow": false,
                    "typescript": true,
                  },
                ],
              ],
            },
          },
        ],
      },
      Object {
        "exclude": Array [
          /node_modules/,
        ],
        "loader": "babel-loader",
        "options": Object {
          "presets": Array [
            Array [
              "/Users/atanasster/component-controls/node_modules/@babel/preset-env/lib/index.js",
              Object {
                "modules": "commonjs",
              },
            ],
            "/Users/atanasster/component-controls/node_modules/@babel/preset-react/lib/index.js",
          ],
        },
        "test": /\\\\\\.\\(md\\|mdx\\)\\$/,
      },
      Object {
        "exclude": Array [
          /node_modules/,
          /node_modules/,
        ],
        "loader": "@component-controls/loader/loader",
        "options": Object {
          "mdx": Object {
            "transformMDX": true,
          },
          "propsLoaders": Array [
            Object {
              "name": "@component-controls/react-docgen-info",
              "test": /\\\\\\.\\(js\\|jsx\\)\\$/,
            },
            Object {
              "name": "@component-controls/react-docgen-typescript-info",
              "test": /\\\\\\.\\(ts\\|tsx\\)\\$/,
            },
          ],
        },
        "test": /\\\\\\.\\(md\\|mdx\\)\\$/,
      },
      Object {
        "enforce": "pre",
        "exclude": Array [
          /node_modules/,
          /node_modules/,
        ],
        "loader": "@component-controls/loader/loader",
        "options": Object {
          "propsLoaders": Array [
            Object {
              "name": "@component-controls/react-docgen-info",
              "test": /\\\\\\.\\(js\\|jsx\\)\\$/,
            },
            Object {
              "name": "@component-controls/react-docgen-typescript-info",
              "test": /\\\\\\.\\(ts\\|tsx\\)\\$/,
            },
          ],
        },
        "test": /\\\\\\.\\(story\\|stories\\)\\.\\(js\\|jsx\\|ts\\|tsx\\)\\$/,
      },
    ],
  },
  "resolve": Object {
    "extensions": Array [
      ".ts",
      ".tsx",
      ".js",
      ".jsx",
    ],
  },
}
`;

exports[`merge instrument and react-docgen-typescript merge options 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "test": /\\\\\\.\\(ts\\|tsx\\)\\$/,
        "use": Array [
          Object {
            "loader": "/Users/atanasster/component-controls/node_modules/babel-loader/lib/index.js",
            "options": Object {
              "presets": Array [
                Array [
                  "react-app",
                  Object {
                    "flow": false,
                    "typescript": true,
                  },
                ],
              ],
            },
          },
        ],
      },
      Object {
        "exclude": Array [
          /node_modules/,
        ],
        "loader": "babel-loader",
        "options": Object {
          "presets": Array [
            Array [
              "/Users/atanasster/component-controls/node_modules/@babel/preset-env/lib/index.js",
              Object {
                "modules": "commonjs",
              },
            ],
            "/Users/atanasster/component-controls/node_modules/@babel/preset-react/lib/index.js",
          ],
        },
        "test": /\\\\\\.\\(md\\|mdx\\)\\$/,
      },
      Object {
        "exclude": Array [
          /node_modules/,
          /node_modules/,
        ],
        "loader": "@component-controls/loader/loader",
        "options": Object {
          "mdx": Object {
            "transformMDX": true,
          },
          "prettier": Object {
            "tabWidth": 4,
          },
          "propsLoaders": Array [
            Object {
              "name": "@component-controls/react-docgen-info",
              "test": /\\\\\\.\\(js\\|jsx\\)\\$/,
            },
            Object {
              "name": "@component-controls/react-docgen-typescript-info",
              "test": /\\\\\\.\\(ts\\|tsx\\)\\$/,
            },
          ],
        },
        "test": /\\\\\\.\\(md\\|mdx\\)\\$/,
      },
      Object {
        "enforce": "pre",
        "exclude": Array [
          /node_modules/,
          /node_modules/,
        ],
        "loader": "@component-controls/loader/loader",
        "options": Object {
          "propsLoaders": Array [
            Object {
              "name": "@component-controls/react-docgen-info",
              "test": /\\\\\\.\\(js\\|jsx\\)\\$/,
            },
            Object {
              "name": "@component-controls/react-docgen-typescript-info",
              "test": /\\\\\\.\\(ts\\|tsx\\)\\$/,
            },
          ],
        },
        "test": /\\\\\\.\\(story\\|stories\\)\\.\\(js\\|jsx\\|ts\\|tsx\\)\\$/,
      },
    ],
  },
  "resolve": Object {
    "extensions": Array [
      ".ts",
      ".tsx",
      ".js",
      ".jsx",
    ],
  },
}
`;
